### ISSUE TITLE
MongoError: document is larger than the maximum size 16777216

### ISSUE THREAD

**Issue by simllll** [Automattic/mongoose#6555]
I want to report a bug / or I have a question.

**What is the current behavior?**
Suddenly I get a lot of "document is larger than the maximum size" errors. Unfortunately I cannot upgrade mongoose due to a session bug where my cursors get killed. So I was not able to test this with the latest verison. I'm on 5.0.14 currently. Anyway, what I get is an uncatchable error like
```
[ERROR] MongoError: document is larger than the maximum size 16777216
    at Function.MongoError.create (/var/node_modules/mongoose/node_modules/mongodb-core/lib/error.js:45:10)
    at toError (/var/node_modules/mongoose/node_modules/mongodb/lib/utils.js:149:22)
    at addToOperationsList (/var/node_modules/mongoose/node_modules/mongodb/lib/bulk/unordered.js:173:11)
    at UnorderedBulkOperation.raw (/var/node_modules/mongoose/node_modules/mongodb/lib/bulk/unordered.js:432:7)
    at bulkWrite (/var/node_modules/mongoose/node_modules/mongodb/lib/collection.js:610:12)
    at executeOperation (/var/node_modules/mongoose/node_modules/mongodb/lib/utils.js:419:24)
    at Collection.insertMany (/var/node_modules/mongoose/node_modules/mongodb/lib/collection.js:512:10)
    at Collection.insert (/var/node_modules/mongoose/node_modules/mongodb/lib/collection.js:784:15)
    at NativeCollection.(anonymous function) [as insert] (/var/node_modules/mongoose/lib/drivers/node-mongodb-native/collection.js:139:28)
    at model.Model.$__handleSave (/var/node_modules/mongoose/lib/model.js:160:21)
    at model.Model.$__save (/var/node_modules/mongoose/lib/model.js:229:8)
    at /var/node_modules/kareem/index.js:276:8
    at _next (/var/node_modules/kareem/index.js:102:16)
    at process.nextTick (/var/node_modules/kareem/index.js:427:38)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickDomainCallback (internal/process/next_tick.js:218:9)
```

It kind of sounds simlar to the already closed bug #3906. But I guess this must be another case, I would like to track it down, but due to the "kareem" hook, I have no usefull backtrace. Is there any chance to see the document that's causing this? It is an "error", not an promise rejection. Which is very strange, cause I actually only have "promises" in my application and all of them are (as far as I know) catched.

**If the current behavior is a bug, please provide the steps to reproduce.**
still working to track it down, need a way to get more debug output.

**What is the expected behavior?**
don't let the process die, call the promise rejection handler correctly.

**Please mention your node.js, mongoose and MongoDB version.**
node js 8.10
mongoose 5.0.14
mongodb 3.6.5

thanks
Simon

**Comment by lineus** [comment]
Hey @simllll! I hope this isn't too wacky of an idea, but what about a pre-save hook that checks the size of the doc? You can do something more meaningful with the info than my example does, but here's what came to mind:

### 6555.js
```js
#!/usr/bin/env node
'use strict';

const assert = require('assert');
const mongoose = require('mongoose');
mongoose.connect('mongodb://localhost/test');
const conn = mongoose.connection;
const Schema = mongoose.Schema;
const sizeof = require('object-sizeof');

let called = 0;
const schema = new Schema({
  data: Buffer
});

schema.pre('save', function() {
  if (sizeof(this) > 16793600) {
    console.log(`${called}: `, this.collection.name, this.id);
    assert.strictEqual(called, rand);
  }
  called++;
});

const Test = mongoose.model('test', schema);

const tests = [];

function big() {
  return Buffer.alloc(17000000, 'wer', 'utf8');
}

let rand = Math.floor(Math.random() * 100);
for (let i = 0; i < 100; i++) {
  let test = (i === rand) ? 
    new Test({ data: big() }):new Test({ data: 'x'});
  tests.push(test);
}

async function run() {
  await conn.dropDatabase();
  await Test.create(tests);
  return conn.close();
}

run().catch(error);

function error(e) {
  console.error(e.errmsg);
  return conn.close();
}
```

### Output:
```
issues: ./6555.js
31:  tests 5b173ab4ed72025ef11d2679
BSONObj size: 17000130 (0x10366C2) is invalid. Size must be between 0 and 16793600(16MB) First element: insert: "tests"
issues:
```

**Comment by vkarpov15** [comment]
@simllll so I looked into this and it looks like this error is a promise rejection, even with a pre save hook. Below script catches the error. Can you modify the below script to repro your issue please?

```javascript
const assert = require('assert');
const mongoose = require('mongoose');

const GITHUB_ISSUE = `gh6555`;
const connectionString = `mongodb://localhost:27017/${ GITHUB_ISSUE }`;
const { Schema } = mongoose;

run().then(() => console.log('done')).catch(error => console.error(error.stack));

async function run() {
  await mongoose.connect(connectionString);
  await mongoose.connection.dropDatabase();
  
  const s = new Schema({ data: Buffer });
  s.pre('save', next => {
    setImmediate(next);
  });
  const Test = mongoose.model('test', s);
  
  const tests = [];
  
  function big() {
    return Buffer.alloc(17000000, 'wer', 'utf8');
  }
  
  let rand = Math.floor(Math.random() * 100);
  for (let i = 0; i < 100; i++) {
    let test = (i === rand) ? 
      new Test({ data: big() }):new Test({ data: 'x'});
    tests.push(test);
  }

  try {
    await Test.create(tests);
  } catch (error) {
    console.log('Caught', error);
  }
  console.log('Done');
}
```

**Comment by simllll** [comment]
Thanks lineus for your idea, Hi vkarpov15, thanks for looking into this. 

TLDR: I was not able to reproduce the issue anymore.

Right now I get following two "cases":
1.) unhandledRejection: Attempt to write outside buffer bounds
stack:
```
RangeError: Attempt to write outside buffer bounds
    at Buffer.write (buffer.js:881:13)
    at serializeString (/var/node_modules/bson/lib/bson/parser/serializer.js:33:14)
    at serializeInto (/var/node_modules/bson/lib/bson/parser/serializer.js:919:17)
    at serializeObject (/var/node_modules/bson/lib/bson/parser/serializer.js:347:18)
    at serializeInto (/var/node_modules/bson/lib/bson/parser/serializer.js:937:17)
    at serializeObject (/var/node_modules/bson/lib/bson/parser/serializer.js:347:18)
    at serializeInto (/var/node_modules/bson/lib/bson/parser/serializer.js:727:17)
    at serializeObject (/var/node_modules/bson/lib/bson/parser/serializer.js:347:18)
    at serializeInto (/var/node_modules/bson/lib/bson/parser/serializer.js:937:17)
    at serializeObject (/var/node_modules/bson/lib/bson/parser/serializer.js:347:18)
    at serializeInto (/var/node_modules/bson/lib/bson/parser/serializer.js:937:17)
    at serializeObject (/var/node_modules/bson/lib/bson/parser/serializer.js:347:18)
    at serializeInto (/var/node_modules/bson/lib/bson/parser/serializer.js:937:17)
    at serializeObject (/var/node_modules/bson/lib/bson/parser/serializer.js:347:18)
    at serializeInto (/var/node_modules/bson/lib/bson/parser/serializer.js:937:17)
    at serializeObject (/var/node_modules/bson/lib/bson/parser/serializer.js:347:18)
    at serializeInto (/var/node_modules/bson/lib/bson/parser/serializer.js:937:17)
    at BSON.serialize (/var/node_modules/bson/lib/bson/bson.js:63:28)
    at Query.toBin (/var/node_modules/mongodb-core/lib/connection/commands.js:144:25)
    at serializeCommands (/var/node_modules/mongodb-core/lib/connection/pool.js:1057:43)
    at Pool.write (/var/node_modules/mongodb-core/lib/connection/pool.js:1273:3)
    at Server.command (/var/node_modules/mongodb-core/lib/topologies/server.js:721:15)
    at Mongos.command (/var/node_modules/mongodb-core/lib/topologies/mongos.js:1105:10)
    at Mongos.command (/var/node_modules/mongodb/lib/topologies/topology_base.js:316:25)
    at executeCommand (/var/node_modules/mongodb/lib/db.js:334:19)
    at executeOperation (/var/node_modules/mongodb/lib/utils.js:420:24)
    at Db.command (/var/node_modules/mongodb/lib/db.js:357:10)
    at findAndModify (/var/node_modules/mongodb/lib/collection.js:2244:13)
    at executeOperation (/var/node_modules/mongodb/lib/utils.js:420:24)
    at Collection.findAndModify (/var/node_modules/mongodb/lib/collection.js:2173:10)
    at findOneAndUpdate (/var/node_modules/mongodb/lib/collection.js:2138:8)
    at executeOperation (/var/node_modules/mongodb/lib/utils.js:420:24)
    at Collection.findOneAndUpdate (/var/node_modules/mongodb/lib/collection.js:2119:10)
    at Agenda.module.exports [as saveJob] (/var/node_modules/agenda/lib/agenda/save-job.js:44:22)
    at Job.module.exports [as save] (/var/node_modules/agenda/lib/job/save.js:9:15)
    at jobCallback (/var/node_modules/agenda/lib/job/run.js:30:14)
    at outboxPromise.then.then (/var/app/current/build/main.js:48503:227)
    at tryCatcher (/var/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/var/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/var/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 (/var/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/var/node_modules/bluebird/js/release/promise.js:693:18)
    at Async._drainQueue (/var/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/var/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/var/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:794:20)
    at tryOnImmediate (timers.js:752:5)
    at processImmediate [as _immediateCallback] (timers.js:729:5)

```
2.) Error: uncaughtException with reason: 
```
{
$clusterTime: {
clusterTime: 6568856925352493357, 
signature: {}
}, 
code: 10334, 
codeName: Location10334, 
errmsg: BSONObj size: 17549583 (0x10BC90F) is invalid. Size must be between 0 and 16793600(16MB) First element: findandmodify: "agendaJobs", 
message: BSONObj size: 17549583 (0x10BC90F) is invalid. Size must be between 0 and 16793600(16MB) First element: findandmodify: "agendaJobs", 
name: MongoError, 
ok: 0, 
operationTime: 6568856925352493357, 
stack: [Filtered]
}
```

Both are kind of related with "agenda" (node js module), could be that there are indeed some cases not covered. So therefore I would say, this is not related to mongoose itself.

The "document is larger than the maximum size" error as described above is solved so far on my side. The issue was related with a Mixed Schema type that had, besides some other references, a large array in it. I shrinked the array down to object ids, and the problem was gone. The problem is, this was only "reproducible" in production and I cannot revert this change anymore.
I tried to create a test script, but as you already stated, it seems like it is a promise rejection and somehow my code didn't pick it up.. 
The weird thing is that I have a custom unhandledRejection and a uncaughtException handler on my process, on my test script the "unhandledRejection" is triggered (as it should be), but on production I'm 100% sure it was the uncaughtException handler (this one kills the process afterwards). Unfortunately I cannot find the log entries anymore. Maybe it was related to something else, I would say we can close this issue for now.

thanks again for looking into this!!

**Comment by vkarpov15** [comment]
Glad you found the issue :+1:
